app/lib/search/updater.dart

```diff
@@ -4,92 +4,107 @@
+        if (isLocalDevMode && counter > 300) break;
+      }
+      if (_firstPackageListCount == null) {
+        _firstPackageListCount = counter;
+        // A bogus task that will be cleared easily and do no harm, and it makes
+        // sure that the first index merge will get called eventually.
+        yield new Task('___placeholder_for_index_trigger___', '');
+      }
+      _lastTs = now.subtract(const Duration(minutes: 10));
+      await new Future.delayed(new Duration(minutes: 30));
     }
   }
+}
 
-  Future _doUpdate(int limit) async {
-    final List<Key> versionKeys = [];
-    final Map<String, String> devVersions = {};
+class BatchIndexUpdater {
+  final List<Task> _batch = [];
+  Timer _batchUpdateTimer;
+  Future _ongoingBatchUpdate;
+  int _taskCount = 0;
 
-    _logger.info('Polling packages for changes.');
-    int count = 0;
-    await for (Package p in dbService.query(Package).run()) {
-      final bool inIndex = await index.contains(
-          _toUrl(p.name), p.latestVersion, p.latestDevVersion);
-      if (inIndex) continue;
-      versionKeys.add(p.latestVersionKey);
-      devVersions[p.name] = p.latestDevVersion;
-      count++;
-      if (limit != null && limit <= count) break;
+  Future updateIndex(Task task) async {
+    await _currentBatchRunning();
+    _batch.add(task);
+    _taskCount++;
+    if (_batch.length < 20) {
+      _batchUpdateTimer ??= new Timer(const Duration(seconds: 10), () {
+        _updateBatch();
+      });
+    } else {
+      await _updateBatch();
     }
-    _logger.info('Found ${versionKeys.length} packages to update.');
-    if (versionKeys.isEmpty) return;
+  }
 
-    for (List<Key> slices in sliceList(versionKeys, 20)) {
-      final Key firstKey = slices.first;
-      final String firstPkg = firstKey.parent.id;
-      final String firstVersion = firstKey.id;
-      _logger.info('Updating packages staring with $firstPkg $firstVersion');
-      final List<PackageVersion> versions = await dbService.lookup(slices);
-      final List<PackageDocument> documents = versions
-          .map((pv) => new PackageDocument(
-                url: _toUrl(pv.package),
-                package: pv.package,
-                version: pv.version,
-                devVersion: devVersions[pv.package],
-                detectedTypes: pv.detectedTypes,
-                description: pv.pubspec.description,
-                lastUpdated: pv.shortCreated,
-                readme: pv.readmeContent,
-                popularity: mockScores[pv.package] ?? 0.0,
-              ))
-          .toList();
-      await index.addAll(documents);
-      _logger.info('Updated ${slices.length} packages.');
-    }
+  Future _updateBatch() async {
+    _batchUpdateTimer?.cancel();
+    _batchUpdateTimer = null;
 
-    _logger.info('Calling index.merge() ...');
-    await index.merge();
-    _logger.info('index.merge() completed.');
+    await _currentBatchRunning();
+    if (_batch.isEmpty) return;
```

Comment:
There is almost a race here (when the timer and `updateTask` will call this function, and both of them get yielded via the `await`). It's quite difficult to see whether it can happen or not, so I think we should simplify it, so it becomes crystal clear there is no race.

It's easier to read if you bundle **checking** and **setting** of `_ongoingBatchUpdate`  (and `_batch`)  together.

Like in
```dart

  while (_ongoingBatchUpdateFuture != null) {
    await _ongoingBatchUpdateFuture;
  }
  final Completer completer = new Completer();
  _ongoingBatchUpdateFuture = completer.future;


  final List<Task> batch = _batch.toList(growable: false);
  _batch.clear();

```